Dynamic memory options

tpihlaja opened the pull request:
Discussion on handling dynamic memory allocations in mbed-client - based on prototype implementation.

tpihlaja committed:
Implement base for new and delete overloads

tpihlaja committed:
Inherit own dynamic memory handling in M2MBase

tpihlaja committed:
Implement alternative init methods and print heap statistics

tpihlaja committed:
Improve compilation flags and statistics prints

tpihlaja committed:
Implement memory allocator in M2MDynamicMemory

tpihlaja committed:
Pass NSDL allocators from M2MDynamicMemory

Mbed-client-c library memory allocations seem to be independent
enough to work with dynmemLIB when rest of code still uses
system malloc. Minor upgrades to M2MDynamicMemory class trace
prints and method signatures. M2MDynmemLIB is enabled.

tpihlaja committed:
Change M2MNsdlInterface to use dynmemLIB

tpihlaja committed:
Change nsdlaccesshelper library to use dynmemLIB

tpihlaja committed:
Change M2MTLVSerializer to use dynmemLIB

tpihlaja committed:
Change M2MBase derivatives to use dynmemLIB

Changed explicit dynamic memory allocations, from classes that inherit
M2MBase, to use memory_alloc and memory_free from M2MDynamicMemory,
instead of system malloc and free.

tpihlaja committed:
Add very basic failure callback to dynmemLIB

tpihlaja committed:
Change some core classes to use dynmemLIB

Changed M2MInterfaceImpl and M2MInterfaceFactory to inherit
M2MDynamicMemory.

tpihlaja committed:
Code clean-up

tpihlaja committed:
Disable prints by default

Commented out trace prints preprocessor flag, improved scope of flags
and removed one unused local variable.

tpihlaja committed:
Really disable prints

tpihlaja commented:
Automatic checks will fail because test applications should initialize dynamic memory somewhere before mbed client classes are used.

tpihlaja committed:
Modify header inclusions to improve portability

tpihlaja committed:
Fix ARM Compiler build

TeroJaasko commented on source/include/m2mdynamicmemory.h:
```diff
@@ -0,0 +1,51 @@
+    void * operator new (size_t size);
+    void operator delete (void * ptr);
+    void * operator new[] (size_t size);
+    void operator delete[] (void * ptr);
```
This system needs also a way to keep the original, libc allocator in place, without touching eg. "operator new" at all.

tpihlaja replied:
Yes, there are no flags in header file currently. I should recheck if module builds without any configuration flag. Just introducing functions shouldn't touch them...?

TeroJaasko commented on source/include/m2mdynamicmemory.h:
```diff
@@ -0,0 +1,51 @@
+    static void init(void *heapAllocation, size_t size);
+    static void print_heap_running_statistics(void);
+    static void print_heap_overall_statistics(void);
+    static void *memory_alloc(uint16_t size);
```
Doesn't this uint16_t cause now a bunch of warnings as the client uses uin32_t and/or size_t on many places?

tpihlaja replied:
yes probably, however we should first make dynmemLIB itself more flexible and build up functionality from there. Now the library seems to work "at 16-bit".

tpihlaja commented on mbed-client/m2mbase.h:
```diff
@@ -48,7 +49,7 @@ class M2MResource;
  *  can be created. This serves base class for Object, ObjectInstances and Resources.
  */
-class M2MBase : public M2MReportObserver {
+class M2MBase : public M2MReportObserver, public M2MDynamicMemory {
```
The namespace overload of new/delete seems to be prohibited by C++ standard. Anyways we could first make namespace changes and then try the overload in different environments. I have a separate ticket for namespace update. Maybe separate namespaces for application and core part. To me inheritance seems more clear way to do "dynamic memory options". It is a bit more brute force and maybe some complications with multiple inheritance might rise, but so far it has worked fine.

tpihlaja commented on source/m2mdynamicmemory.cpp:
```diff
@@ -0,0 +1,249 @@
+#endif
+static void memory_fail_callback(heap_fail_t fail) {
+#ifdef M2M_TRACE_PRINTS
+    printf("\nM2M memory failure: %u\n", fail);
```
These can be considered temporary debug prints. I should study the mbed_trace system a bit before changing to use it.

tpihlaja commented on source/m2mnsdlinterface.cpp:
```diff
@@ -74,7 +74,7 @@ M2MNsdlInterface::M2MNsdlInterface(M2MNsdlObserver &observer, M2MConnectionHandl
     // and receiving purposes.
     _nsdl_handle = sn_nsdl_init(&(__nsdl_c_send_to_server), &(__nsdl_c_received_from_server),
-                 &(__nsdl_c_memory_alloc), &(__nsdl_c_memory_free));
+                 &(memory_alloc), &(memory_free));
```
Allright. When introducing new wrappers I tried to get rid of as many old ones as possible.

tpihlaja replied:
In fact signatures seemed to be same:
-void* __nsdl_c_memory_alloc(uint16_t size)

Both used uint16_t type size parameter.

tpihlaja commented on source/m2mdynamicmemory.cpp:
```diff
@@ -0,0 +1,249 @@
+    void *tmp;
+    tmp=m2m_dyn_mem_alloc(size);
+#ifdef M2M_TRACE_PRINTS
+    memTotal+=size; memCount++;
```
Ok.

tpihlaja committed:
Keep book keeping variables in heap sector header

tpihlaja committed:
Add handle to dynmemLIB interface

Use heap sector address also as handle to heap instance.

tpihlaja committed:
Implement basic sector header book keeping

tpihlaja committed:
Expand book keeping header

Added failure callback and memory status information pointers to
heap sector book keeping header.

tpihlaja committed:
Clean-up code

tpihlaja committed:
Create heap classes as copies

Created two copies of M2MDynamicMemory heap class:
 * M2MCoreMemory, to be used by core part of mbed-client
 * M2MApplicationMemory, to be used by application part of mbed-client

tpihlaja committed:
Use two heaps in mbed-client

tpihlaja committed:
Make m2m dynmemLIB symbols unique

tpihlaja committed:
Improve preprocessor flags

tpihlaja committed:
Delete obsolete M2MDynamicMemory class

kjbracey-arm commented:
I've assume this will be properly integrated with the existing ns_dynmemlib, and you're not really intending to add a second heap manager to the build?

yogpan01 commented:
Agreed with @kjbracey-arm , this extension should be integrated with existing dynmemlib so that other components can also utilize this feature.
@tpihlaja FYI. To begin with this new component can be part of nanostack-libservice , if directly integrating into existing dynmemlib is too big task at the moment.

kjbracey-arm commented:
I would assume the existing nsdynmemlib API would be easily mapped to to the updated implementation by just passing a fixed heap handle.

And then there should be some way to get that base handle for the case where your system does want mbed-client or other components to share the "main" dynmemlib heap, rather than create a separate one.

tpihlaja committed:
Move M2MTLVSerializer to M2MCoreMemory space

tpihlaja committed:
Free buffer allocated by String with system free

tpihlaja committed:
Copy Get response payload to M2MCoreMemory space

tpihlaja committed:
Restore nsdl_c allocator-wrappers

tpihlaja committed:
Change m2m dynmemLIB interface to use size_t type

kjbracey-arm commented:
On the smallest platforms, it can be useful to restrict to 16 bits, even if the architecture in principle supports bigger. MSP430 is an example - it has 32-bit size_t, but we don't want/need to use that, and we know our heap(s) will be <64K.  Using size_t everywhere for all object sizes can be inefficient from a RAM point of view.

We probably want a typedef for the internal size type, to leave that possibility open. Note that presumably the internal heap implementation should use that too - as I recall it's using just "int", so this current API be giving 32-bit API but 16-bit heap on the MSP430.

There's also the potential possibility for heap objects being <64K but the heap total being >64K, which I don't think has been properly thought about.  You might want different types for heap size and individual object size.

In standard C, size_t is the appropriate size for an object, and there's no type to describe the size of a heap. In this sort of embedded heap, actually size_t is the appropriate size for the heap, but the size for the objects could be smaller.

tpihlaja commented:
Inner working of dynmemLIB definitely needs some work to support latest size_t API change. 64K maximum heap size sounds small to me. If that is the fact then we should keep the original implementation. With size_t we aimed for generic implementation. I guess we could still (with size_t in API) limit heap and object size somehow to keep inner working simpler and more memory efficient.

kjbracey-arm commented:
Using size_t is generic, but excessively fat for some of our embedded targets. Having your own specific typedef allows a build-time choice independent of the system.

We certainly want to go above 64K total for the heap - I'm not too concerned about making the heap total stats APIs 32-bit, and I thought that had already been done somewhat. So I'd be okay with those numbers just being size_t.

But for the individual heap objects, so far they've always been way less than 64K, so having a typedef for the object size which is used both in APIs and in the internal heap structure would maintain existing flexibility - we could define it as smaller than size_t on teeny platforms.

And that size type could be maybe exposed in higher application APIs too - there's a whole bunch of fixed uint16_ts in APIs that I'd like to see explanded, but I wouldn't want them to be forced to size_t on every platform.

tpihlaja committed:
Use custom typedef for block size in dynmemLIB API

tpihlaja committed:
Use custom typedef for internal block size

External API type is now the original uint16_t, and internal is the
original int.

tpihlaja committed:
Change internal block size data type to int16_t

tpihlaja committed:
Add custom type for heap total size

tpihlaja committed:
Fix compilation without new/delete overloads

Rearrange preprocessor flags and create unique flags for each heap
class. It should now be possible to configure heaps completely
independently and also include header files of different heap classes
in same compilation unit.

tpihlaja committed:
Update unit tests

Added stubs for new classes and added the new stubs to CMakeLists
where needed. Removed some memory allocator tests where
implementation was removed due to common implementation in
the new classes. Now unit tests pass if new/delete overloads
are disabled.

tpihlaja committed:
Use m2mdynmemLIB from nanostack-libservice

tpihlaja committed:
Use mbed-trace in heap classes

tpihlaja committed:
Clean-up heap classes

tpihlaja committed:
Clean-up heap classes more

 * Rearranged pre-processor flags.

 * Removed unused statistics variables from dynmemLIB configuration.

 * Made dynmemLIB memory failure callback print unique to heap.
